pygpu/gpuarray.pyx

```diff
@@ -1465,6 +1465,57 @@ cdef class GpuArray:
         else:
             raise IndexError, "cannot index with: %s" % (key,)
 
+    def write(self, np.ndarray src not None):
+        """Writes host's Numpy array to device's GpuArray.
+
+        :param src: source array in host
+        :type src: np.ndarray
+
+        :raises ValueError: If this GpuArray is not compatible with `src`
+
+        """
+        if not self.flags.behaved:
+            raise ValueError, "Destination GpuArray is not well behaved: aligned and writeable"
+        if not ((self.flags.c_contiguous and src.flags['C_CONTIGUOUS'] and src.flags['ALIGNED']) or \
+                (self.flags.f_contiguous and src.flags['F_CONTIGUOUS'] and src.flags['ALIGNED'])):
+            raise ValueError, "GpuArray and Numpy array do not match in contiguity or Numpy array is not aligned"
+        if self.dtype != src.dtype:
+            raise ValueError, "GpuArray and Numpy array do not have matching data types"
+        cdef size_t npsz = np.PyArray_NBYTES(src)
+        cdef size_t sz = gpuarray_get_elsize(self.ga.typecode)
+        cdef unsigned i
+        for i in range(self.ga.nd):
+            sz *= self.ga.dimensions[i]
+        if sz != npsz:
+            raise ValueError, "GpuArray and Numpy array do not have the same size in bytes"
+        array_write(self, np.PyArray_DATA(src), sz)
+
+    def read(self, np.ndarray dst not None):
+        """Reads from this GpuArray into host's Numpy array.
+
+        :param dst: destination array in host
+        :type dst: np.ndarray
+
+        :raises ValueError: If this GpuArray is not compatible with `src`
+
```

Comment:
I would add in the docstring why this is useful as otherwise people can simply use a_gpu_ndarray[...] = an_ndarray and it will work and it could be optimized to be as efficient as the current case.

The difference is mainly that this one only check the total number of element and not the shapes. Also tell that this will always be as fast or faster then  a_gpu_ndarray[...] = an_ndarray.